### PR TITLE
Fix linear API in models with slack

### PR DESCRIPTION
**Pull request by tmigot**
I added a test of the linear API in #57 , which helps detect and debug this.

### CODE REVIEW

File: src/slack-model.jl [hunk 1]
```diff
@@ -1,6 +1,29 @@
...
+end
+
+"""
+    get_relative_indices(model)
+
+Return two sets of triplet with the relative indices of `jlow`, `jupp` and `jrng`
+within the set of linear and nonlinear indices.
+"""
+function get_relative_indices(model)
+  lin = model.meta.lin
```
Comment (dpo) [inline src/slack-model.jl:17]: You're not saving much by defining this variable since you still write `model.meta.jlow`, etc., below. I suggest removing this variable.

File: src/slack-model.jl [hunk 2]
```diff
@@ -1,6 +1,29 @@
...
+  if intersect(jl, ind) == []
+    return Int[]
+  end
+  return map(b -> findfirst(x -> x == b, ind), intersect(jl, ind))
+end
+
+"""
+    get_relative_indices(model)
+
+Return two sets of triplet with the relative indices of `jlow`, `jupp` and `jrng`
```
Comment (dpo) [inline src/slack-model.jl:13]:
```suggestion
Return two sets of triplets with the relative indices of `jlow`, `jupp` and `jrng`
```

Comment (dpo) [reply src/slack-model.jl:13]: NB: this function doesn't return two sets of triplets. A set of triplets would be `[(a, b, c), (d, e, f), ...]`.

File: src/slack-model.jl [hunk 3]
```diff
@@ -425,15 +459,18 @@ function NLPModels.jtprod_nln!(
...
   nlow, nupp, nrng = length(jlow), length(jupp), length(jrng)
+  nslacklin = length(nlp.jlow_lin) + length(nlp.jupp_lin) + length(nlp.jrng_lin)
   @views begin
     jtprod_nln!(nlp.model, x[1:n], v, jtv[1:n])
+    jtv[(n + 1):(n + nslacklin)] .= zero(T)
+    n += nslacklin
     jtv[(n + 1):(n + nlow)] = -v[jlow]
     jtv[(n + nlow + 1):(n + nlow + nupp)] = -v[jupp]
     jtv[(n + nlow + nupp + 1):(n + nlow + nupp + nrng)] = -v[jrng]
-    jtv[(n + nlow + nupp + nrng + 1):(nlp.meta.nvar)] .= zero(T)
```
Comment (dpo) [inline src/slack-model.jl:436]: Are there dots missing above?

**Review by dpo — COMMENTED** [review]
This one is hard to check...

File: src/slack-model.jl [hunk 4]
```diff
@@ -1,6 +1,29 @@
 export SlackModel, SlackNLSModel
 
-get_slack_ind(jl, ind) = map(b -> findfirst(x -> x == b, ind), intersect(jl, ind))
+function get_slack_ind(jl, ind)
```
[the remark this answers is not in the document]
Comment (tmigot) [reply src/slack-model.jl:8]: It returns the indices of `jl` relative to the set `ind`. So, [`2`] within [`2`, `4`] is [`1`]. I added a docstring saying `Return the relative indices of the set of indices jl within the set of indices ind.`

File: src/slack-model.jl (hunk 2 above)
Comment (tmigot) [reply src/slack-model.jl]: Thanks for the suggestions. I removed `two sets of triplets with`

File: src/slack-model.jl (hunk 3 above)
Comment (tmigot) [reply src/slack-model.jl:436]: Thanks for catching that.

File: src/slack-model.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (dpo) [reply src/slack-model.jl:8]: It sounds like what `findall` does?

[the remark this answers is not in the document]
Comment (tmigot) [reply src/slack-model.jl:8]: Maybe, I am not sure to see how that would work though

[the remark this answers is not in the document]
Comment (abelsiqueira) [reply src/slack-model.jl:8]: I think you want `indexin(jl, ind)`. It returns `[findfirst(j .== ind) for j in jl]`. The difference is the `intersect`, which I think you don't need, unless you expect repetition. If that is the case, you can call `unique` on the result.

[the remark this answers is not in the document]
Comment (tmigot) [reply src/slack-model.jl:8]: Thanks for the suggestion! I need the interesect because some elements of `jl` might not be in `ind`. I added:
```
get_slack_ind(jl, ind) = findall(!isnothing, indexin(jl, ind))
```
and added the `findall` because `indexin` return a `Vector{Union{Int, Nothing}}`.

**Review by abelsiqueira — APPROVED** [review]
It is a bit hard to follow, so I hope that #57 makes it easier to see that it is not breaking. Thanks for the PR
